feat: LineChart and ChartGrid multi-use components [WEB-659] [WEB-685]

mapmeld opened the pull request:
## Description

LineChart as a reusable component (to be tested in UI Kit).

Current situation:
- [x] basic labeling and dimensions of chart
- [x] data accepts array of lines, which are arrays of `[x, y]` points.
- [x] tooltip, legend plugins can be enabled
- [x] can set and update focusedSeries
- [x] ChartGrid displays an array of chart info in a grid, using react-window's grid
- [x] ChartGrid uses SyncableBounds to share cursor and zoom

Todos (given to future tickets):
- closestPoint plugin
- UI to select metrics within a chart or chartgrid
- x-axis should be synced within a chartgrid

## Test Plan

There are two main things which need to get checked at this point:

- reasonable API / parameters for future use of this component
- `/det/design` displays charts and a chartgrid demo-ing the labeling options, plotting data, grid of charts, etc.

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

ClaireNeveu commented on webui/react/src/pages/DesignKit.tsx:
```diff
@@ -331,6 +334,65 @@ const DropdownsSection: React.FC = () => {
+        />
+      </Card>
+      <Card title="Tooltip options">
+        <p>A chart with a tooltip.</p>
```
Charts should always have a tooltip.

mapmeld replied:
fixed in commit

julian-determined-ai commented:
@mapmeld question: looking at the ticket we proposed certain other features such as shared zoom, adding/removing charts,  etc that I dont see noted in this PR. Is the plan to support those features in other tickets also?

ClaireNeveu commented on webui/react/src/components/kit/LineChart.tsx:
```diff
@@ -0,0 +1,231 @@
+interface Props {
+  focusedSeries?: number;
+  height?: number;
+  onSeriesHover?: (seriesIdx: number | null) => void;
```
This feels like more arguments than we actually need. It's not clear to me when we need most of these callbacks.

We want to avoid fixed width and standardize whether metrics are displayed, whether legends are displayed, etc.

Which of these are actually necessary

mapmeld replied:
✅ I can remove width from chart since that was needed only for `<AutoSizer>` -> `<FixedSizeGrid>`
I'll re-read the ticket re: callbacks or we can err on not having so many.  If the component can add/select/remove metrics on its own then we would need callbacks.

ClaireNeveu replied:
I think that's only the case if you want to make it a controlled component, no? If the chart manages its own state you shouldn't need any callbacks.

mapmeld replied:
ok I drop the callbacks in this commit.  I think we have a component in multi-trial search where we currently have with chart interactions limit which trials show up in a table, but I might just remembering the parallel coordinates plot

mapmeld commented:
@julian-determined-ai @ClaireNeveu I updated the zoom setup. In the ChartGrid component at the end, there is a sync between the two charts' cursor and zoom.

I've updated the settings so you can only select-zoom along the x-axis.

After some testing, when you select a range (say 5.0 to 10.0) then you're selecting only the points inside that range.  If a chart has no points within the range, then the y-axis will turn blank on that plot, no lines visible.  I added a few points to make valid zooms more common.  In practice we often have many points on the plot and in a similar x-axis range.

trentwatt commented on webui/react/src/components/kit/LineChart.tsx:
```diff
@@ -0,0 +1,172 @@
+  }, [series]);
+
+  const chartOptions: Options = useMemo(() => {
+    const plugins = [tooltipsPlugin({ isShownEmptyVal: false })];
```
@alexjdetermined was wanting the tooltip to show all the series, not just the hovered one(s).

Further enhancement: highlighting the hovered series in the tooltip.

mapmeld replied:
if you have a series A with points at {1,3,5} and series B with points at {0,1,2,3,4,5}
`isShownEmptyVal` is filtering out `null`s, such as value of series A at x=2

the new tooltip (I might want to make it a separate file while we are switching over) does sort the series in it from highest y-value to lowest, and bolds the closest ones (along the y-axis)

<img width="360" alt="Screen Shot 2023-01-18 at 12 55 39 PM" src="https://user-images.githubusercontent.com/643918/213281381-3759d9a0-4cc2-4c24-9390-24daf682d61e.png">

mapmeld commented on webui/react/src/components/kit/LineChart.tsx:
```diff
@@ -0,0 +1,172 @@
+
+interface Serie {
+  color?: string;
+  data: number[][];
```
I was sending it in as [[x1, y1], [x2, y2]]. Between this and data prep, how would the data be input and formatted for the chart?

ClaireNeveu replied:
Sounds like you want `Array<[number, number]>` then (you're passing a tuple).

julian-determined-ai reviewed (approved):
LGTM after open feedback is responded to, and with the understanding further iteration on the components will occur.
